core/src/smartcontracts/isi/asset.rs

```diff
@@ -54,89 +54,165 @@ pub mod isi {
+            }
+        };
+    }
+
+    impl_asset_instruction_info!(u32, AssetValueType::Quantity, 0_u32);
+    impl_asset_instruction_info!(u128, AssetValueType::BigQuantity, 0_u128);
+    impl_asset_instruction_info!(Fixed, AssetValueType::Fixed, Fixed::ZERO);
+
+    /// Trait for blanket mint implementation.
+    trait InnerMint {
+        fn execute<W: WorldTrait, Err>(
+            mint: Mint<Asset, Self>,
             _authority: <Account as Identifiable>::Id,
             wsv: &WorldStateView<W>,
-        ) -> Result<(), Self::Error> {
-            let asset_id = self.destination_id;
-
-            assert_can_mint(&asset_id.definition_id, wsv, AssetValueType::Fixed)?;
-            wsv.asset_or_insert(&asset_id, Fixed::ZERO)?;
+        ) -> Result<(), Err>
+        where
+            Self: AssetInstructionInfo + TryCheckedAdd + FloatMetricFrom + Copy,
+            AssetValue: From<Self> + TryAsMut<Self>,
+            Value: From<Self>,
+            <AssetValue as TryAsMut<Self>>::Error: std::error::Error + Send + Sync + 'static,
+            Err: From<Error>,
+        {
+            let asset_id = mint.destination_id;
+
+            assert_can_mint(
+                &asset_id.definition_id,
+                wsv,
+                <Self as AssetInstructionInfo>::EXPECTED_VALUE_TYPE,
+            )?;
+            wsv.asset_or_insert(
+                &asset_id,
+                <Self as AssetInstructionInfo>::DEFAULT_ASSET_VALUE,
+            )?;
             wsv.modify_asset(&asset_id, |asset| {
-                let quantity: &mut Fixed = asset
+                let quantity: &mut Self = asset
                     .try_as_mut()
                     .map_err(eyre::Error::from)
                     .map_err(|e| Error::Conversion(e.to_string()))?;
-                *quantity = quantity.checked_add(self.object)?;
-                wsv.metrics.tx_amounts.observe((*quantity).into());
+                *quantity = quantity.try_checked_add(mint.object)?;
+                wsv.metrics
+                    .tx_amounts
+                    .observe(<Self as FloatMetricFrom>::from(*quantity));
 
                 Ok(AssetEvent::Added(asset_id.clone()))
-            })
+            })?;
+            Ok(())
         }
     }
 
+    macro_rules! impl_mint {
+        ($ty:ty, $error:ty, $metrics:literal) => {
+            impl InnerMint for $ty {}
+
+            impl<W: WorldTrait> Execute<W> for Mint<Asset, $ty> {
+                type Error = Error;
+
+                #[metrics(+$metrics)]
+                fn execute(
+                    self,
+                    authority: AccountId,
+                    wsv: &WorldStateView<W>,
+                ) -> Result<(), Self::Error> {
+                    <$ty as InnerMint>::execute(self, authority, wsv)
+                }
+            }
+        };
+    }
+
+    impl_mint!(u32, Error, "mint_qty");
+    impl_mint!(u128, Error, "mint_big_qty");
+    impl_mint!(Fixed, Error, "mint_fixed");
```

Comment:
```suggestion
    impl_mint_asset!(u32, u128, Fixed);
```

I think you can go with the alternative design.
The metric names can be removed by having some map e.g. from `u32` to `"qty"`

```rust
impl Value {
    //
    fn metric_name(&self) -> &'static str {
        //
    }
}
```

Reply:
Thank you for suggestion, do you have some kind of example for this? All my attempts to implement such design have failed, `#[metrics(+"metric_name")]` attribute requires literal value.

Reply:
I found that I haven't found any implementation for that. Have you considered changing the `metrics` attribute macro?

Reply:
If you don't have any experience with `proc` macros, you can leave it as is and create a ticket to fix it. 

With that said, three lines vs. one line is still readable to me.

Reply:
I will create a ticket after merge.